### PR TITLE
Use GNU++17

### PR DESCRIPTION
**Pull request by 3d-gussner**
Needed for error conversion code in Prusa-Firmware

**!!! Merging order !!!**

- [x] This PR has to be merged before
  - [x]  Prusa-Firmware [PR](https://github.com/prusa3d/Prusa-Firmware/pull/3500)

!!! To Do !!!
- [ ] Before error conversion code we need to update
  - [x] `PF-build.sh`
  - [ ] the build environment for `build.sh` / travis build
  - [ ] the build server

PFW-1321

### CODE REVIEW

**Review by DRracer — CHANGES_REQUESTED** [review]
please change:
`compiler.ldflags=`
to
`compiler.ldflags=-flto -fwhole-program`

**Comment by wavexx** [comment]
I remember right that -flto + -fwhole-program doesn't make sense. From the gcc 10 documentation:

```
   -fwhole-program
          <snip>
           This option should not be used in combination with -flto.  Instead
           relying on a linker plugin should provide safer and more precise
           information.
```

I would remove -fwhole-program in this case.

File: IDE_Board_Manager/prusa3dboards/platform.txt [hunk 1]
```diff
@@ -25,14 +25,14 @@ compiler.c.elf.flags=-Os -g -flto -Wl,--gc-sections
...
-compiler.cpp.flags=-c -g -Os {compiler.warning_flags} -std=gnu++11 -ffunction-sections -fdata-sections -MMD -flto -fno-fat-lto-objects -fno-exceptions -fno-threadsafe-statics -fno-rtti
+compiler.cpp.flags=-c -g -Os {compiler.warning_flags} -std=gnu++17 -ffunction-sections -fdata-sections -MMD -flto -fno-fat-lto-objects -fno-exceptions -fno-threadsafe-statics -fno-rtti
 compiler.ar.cmd=avr-gcc-ar
 compiler.ar.flags=rcs
 compiler.objcopy.cmd=avr-objcopy
 compiler.objcopy.eep.flags=-O ihex -j .eeprom --set-section-flags=.eeprom=alloc,load --no-change-warnings --change-section-lma .eeprom=0
 compiler.elf2hex.flags=-O ihex -R .eeprom
 compiler.elf2hex.cmd=avr-objcopy
-compiler.ldflags=
+compiler.ldflags=-flto -fwhole-program
```
Comment (wavexx) [inline IDE_Board_Manager/prusa3dboards/platform.txt]: Just use -flto here. -fwhole-program and -flto perform the same duty, but work with different principles. -fwhole-program would require to compile and link in a single pass, which is not technically possible using arduino-builder anyway.

We are already using -flto in the compilation flags, so everything looks correct to me.

File: IDE_Board_Manager/prusa3dboards/platform.txt [hunk 2]
```diff
@@ -25,14 +25,14 @@ compiler.c.elf.flags=-Os -g -flto -Wl,--gc-sections
 compiler.c.elf.cmd=avr-gcc
 compiler.S.flags=-c -g -x assembler-with-cpp -flto -MMD
 compiler.cpp.cmd=avr-g++
-compiler.cpp.flags=-c -g -Os {compiler.warning_flags} -std=gnu++11 -ffunction-sections -fdata-sections -MMD -flto -fno-fat-lto-objects -fno-exceptions -fno-threadsafe-statics -fno-rtti
+compiler.cpp.flags=-c -g -Os {compiler.warning_flags} -std=gnu++17 -ffunction-sections -fdata-sections -MMD -flto -fno-fat-lto-objects -fno-exceptions -fno-threadsafe-statics -fno-rtti
 compiler.ar.cmd=avr-gcc-ar
```
Comment (wavexx) [inline IDE_Board_Manager/prusa3dboards/platform.txt:29]: I cannot add a comment on the exact line above, but on ``compiler.c.flags`` I would also set -std=gnu17 to match the same standard for both c and c++.

**Review by wavexx — CHANGES_REQUESTED** [review]

File: IDE_Board_Manager/prusa3dboards/platform.txt (hunk 2 above)
Comment (leptun) [reply IDE_Board_Manager/prusa3dboards/platform.txt:29]: I tried that and I think that is not possible

Comment (wavexx) [reply IDE_Board_Manager/prusa3dboards/platform.txt:29]: Yeah, just checked now. Looks like gcc7 doesn't support it. Well then, nothing we can do about this last point.

File: IDE_Board_Manager/prusa3dboards/platform.txt (hunk 1 above)
Comment (3d-gussner) [reply IDE_Board_Manager/prusa3dboards/platform.txt]: @DRracer Please comment on that.

Comment (DRracer) [reply IDE_Board_Manager/prusa3dboards/platform.txt]: ok, I agree, let's keep just -flto in ldflags

**Review by wavexx — APPROVED** [review]

**Review by DRracer — APPROVED** [review]

**Pull request merged**